File: app/src/main/java/de/westnordost/osmagent/data/QuestDownloader.java
package de.westnordost.osmagent.data;

import android.content.SharedPreferences;
import android.util.Log;

import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;

import javax.inject.Inject;
import javax.inject.Provider;

import de.westnordost.osmagent.Prefs;
import de.westnordost.osmagent.data.osm.download.OsmQuestDownload;
import de.westnordost.osmagent.data.osmnotes.OsmNotesDownload;
import de.westnordost.osmapi.map.data.BoundingBox;
import de.westnordost.osmapi.map.data.LatLon;

public class QuestDownloader
{
	private static final String TAG = "QuestDownload";

	private final SharedPreferences prefs;
	private final Provider<OsmNotesDownload> notesDownloadProvider;
	private final Provider<OsmQuestDownload> questDownloadProvider;

	private ExecutorService executorService;

	private AtomicBoolean cancelState;

	private VisibleQuestListener listener;

	@Inject public QuestDownloader(SharedPreferences prefs,
								   Provider<OsmNotesDownload> notesDownloadProvider,
								   Provider<OsmQuestDownload> questDownloadProvider)
	{
		this.prefs = prefs;
		this.notesDownloadProvider = notesDownloadProvider;
		this.questDownloadProvider = questDownloadProvider;

		executorService = Executors.newSingleThreadExecutor();
	}

	public void cancel()
	{
		if(cancelState != null)
		{
			cancelState.set(true);
		}
	}

	public void setQuestListener(VisibleQuestListener listener)
	{
		this.listener = listener;
	}

	public void download(final BoundingBox bbox, final Integer maxVisibleQuests)
	{
		cancel();
		cancelState = new AtomicBoolean(false);

		executorService.submit(
				new Runnable()
				{
					@Override public void run()
					{
						if(cancelState.get()) return;

						Set<LatLon> notesPositions = null;
						OsmNotesDownload notesDownload = notesDownloadProvider.get();
						notesDownload.setQuestListener(listener);

						Long userId = prefs.getLong(Prefs.OSM_USER_ID, -1);
						if(userId == -1) userId = null;

						try
						{
							int maxNotes = maxVisibleQuests != null ? maxVisibleQuests : 10000;
							notesPositions = notesDownload.download(bbox, userId, maxNotes);
						}
						catch(Exception e)
						{
							Log.e(TAG, "Unable to download notes", e);
						}

						Integer maxOsmQuestsToRetrieve = maxVisibleQuests;
						if(maxOsmQuestsToRetrieve != null)
						{
							maxOsmQuestsToRetrieve -= notesDownload.getVisibleQuestsRetrieved();
						}

						OsmQuestDownload questDownload = questDownloadProvider.get();
						questDownload.setQuestListener(listener);

						try
						{
							questDownload.download(bbox, notesPositions, maxOsmQuestsToRetrieve, cancelState);
						}
						catch(Exception e)
						{
							Log.e(TAG, "Unable to download osm quests", e);
						}
					}
				}
		);
	}

	public void shutdown()
	{
		cancel();
		executorService.shutdown();
	}

}

File: app/src/main/java/de/westnordost/streetcomplete/data/download/QuestDownloader.kt
package de.westnordost.streetcomplete.data.download

import android.util.Log
import de.westnordost.osmapi.map.data.BoundingBox
import de.westnordost.streetcomplete.data.osmnotes.OsmNotesDownloader
import de.westnordost.streetcomplete.data.osm.osmquest.*
import de.westnordost.streetcomplete.data.osm.osmquest.OsmApiQuestDownloader
import de.westnordost.streetcomplete.data.quest.QuestTypeRegistry
import de.westnordost.streetcomplete.data.user.UserStore
import java.util.concurrent.atomic.AtomicBoolean
import javax.inject.Inject
import javax.inject.Provider

/** Takes care of downloading all note and osm quests */
class QuestDownloader @Inject constructor(
    private val osmNotesDownloaderProvider: Provider<OsmNotesDownloader>,
    private val osmApiQuestDownloaderProvider: Provider<OsmApiQuestDownloader>,
    private val questTypeRegistry: QuestTypeRegistry,
    private val userStore: UserStore
) : Downloader {

    @Synchronized override fun download(bbox: BoundingBox, cancelState: AtomicBoolean) {
        if (cancelState.get()) return

        Log.i(TAG, "(${bbox.asLeftBottomRightTopString}) Starting")

        try {
            downloadQuestTypes(bbox, cancelState)
        } finally {
            Log.i(TAG, "(${bbox.asLeftBottomRightTopString}) Finished")
        }
    }

    private fun downloadQuestTypes(bbox: BoundingBox, cancelState: AtomicBoolean) {
        // always first download notes, note positions are blockers for creating other quests
        downloadNotes(bbox)

        if (cancelState.get()) return

        downloadOsmElementQuestTypes(bbox)
    }

    private fun downloadNotes(bbox: BoundingBox) {
        val notesDownload = osmNotesDownloaderProvider.get()
        val userId: Long = userStore.userId.takeIf { it != -1L } ?: return
        // do not download notes if not logged in because notes shall only be downloaded if logged in

        val maxNotes = 10000
        notesDownload.download(bbox, userId, maxNotes)
    }

    private fun downloadOsmElementQuestTypes(bbox: BoundingBox) {
        val questTypes = questTypeRegistry.all.filterIsInstance<OsmElementQuestType<*>>()
        osmApiQuestDownloaderProvider.get().download(questTypes, bbox)
    }

    companion object {
        private const val TAG = "QuestDownload"
    }
}
